app/models/immunisation_import.rb

```diff
@@ -68,6 +68,13 @@ def process!(patient_session:)
     parse_rows! if rows.nil?
     return if invalid?
 
+    rows
+      .map(&:to_location)
+      .uniq(&:urn)
+      .reject(&:invalid?)
+      .map(&:attributes)
+      .tap { Location.create! _1 }
+
```

Comment:
If I read this right, the locations are created separately from the vaccination records, i.e. all the locations are created first and then all the vaccination records. This feels a little counter-intuitive to me, although maybe it's fine really, except that if one of the vaccination records can't be created then I think the locations and all the vaccination records up until that one stay. So a couple questions:

- Should we wrap this up as a transaction?
- If not, are we happy to have all the locations created? Thinking of a case where the vaccination record causes an error, maybe because it's completely wrong, then there's a chance the associated location shouldn't have been created?
- Do we have a sense of how many different locations we would expect to see in an uploaded file? This might be moot if it should only ever be one, but then, mistakes do happen and maybe we should check that.

Reply:
You're right, wrapped in a transaction. That's the simplest solution for now. We should revisit this and make the validation/error recovery more in-depth as necessary